### PR TITLE
Implement HTTP2 lightyServerBuider

### PR DESCRIPTION
**Pull request by PeterSuna**
- Create and implement HTTP2 lightyServerBuider to RNC app.
- Add tests

Rebased on PR #1009
ID: 1819

### CODE REVIEW

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+                && config.getKeyStorePassword().equals(defaultConfig.getKeyStorePassword())
+                && config.getKeyStoreType().equals(defaultConfig.getKeyStoreType());
+    }
+
+    private static KeyStore getCustomKeystore(final String ksPath, final String ksType, final String ksPassword)
+            throws ConfigurationException, IOException, KeyStoreException, CertificateException,
+            NoSuchAlgorithmException {
+        LOG.info("Loading custom keystore at path : {}", ksPath);
+        final PasswordProtection passProtection = new PasswordProtection(ksPassword.toCharArray());
+        final KeyStore keyStore = Builder.newInstance(ksType, null, passProtection).getKeyStore();
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java:209]: ignore

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            final Path parent = Paths.get(path).getParent();
+            if (!parent.toFile().exists()) {
+                Files.createDirectories(parent);
+            }
+            final KeyPair keyPair = createKeyPair();
+            final X509Certificate certificate = getSelfSignCert(keyPair);
+            keyStore.load(null, ksPassword.toCharArray());
+            keyStore.setKeyEntry("certAlias", keyPair.getPrivate(), ksPassword.toCharArray(),
+                    new X509Certificate[] { certificate });
+            try (FileOutputStream fileOutputStream = new FileOutputStream(path)) {
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java:148]: ignore

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            // if exists, load
+            LOG.info("Loading default keystore at path : {}", resource.getPath());
+            try (FileInputStream fileInputStream = new FileInputStream(resource.getFile())) {
+                keyStore.load(fileInputStream, ksPassword.toCharArray());
+            }
+        } else {
+            // if not exists, create
+            final String path = LightyServerConfigUtils.class.getResource("/").getPath() + ksPath;
+            LOG.info("Creating default keystore at path : {}", path);
+            final Path parent = Paths.get(path).getParent();
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java:139]: ignore

File: lighty-modules/lighty-jetty-server/pom.xml [hunk 4]
```diff
@@ -22,6 +22,10 @@
     <url>https://github.com/PANTHEONtech/lighty</url>
 
     <dependencies>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/pom.xml:25]: ignore

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory;
+import org.eclipse.jetty.http.HttpVersion;
+import org.eclipse.jetty.http2.server.HTTP2ServerConnectionFactory;
+import org.eclipse.jetty.server.HttpConfiguration;
+import org.eclipse.jetty.server.HttpConnectionFactory;
+import org.eclipse.jetty.server.SecureRequestCustomizer;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+
+public class Http2LightyServerBuilder  extends LightyServerBuilder {
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java:21]: Change `Http2LightyServerBuilder  extends` to `Http2LightyServerBuilder extends`.

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    private final SecurityConfig securityConfig;
+
+    public Http2LightyServerBuilder(final InetSocketAddress inetSocketAddress, final SecurityConfig securityConfig) {
+        super(inetSocketAddress);
+        this.server = new Server();
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java:27]: We are using `build()` to create an instance of `Server`. Remove `this.server = new Server();` form here.
This class does not look like builder at all - everything is set in constructor - maybe remove constructor and add methods `withAddress(InetSocketAddress)` and `withSecurityConfig(SecurityConfig)`.

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java [hunk 7]
```diff
@@ -108,6 +108,7 @@ private static Optional<InputStream> readKeyStoreFile(final String keyStoreFileP
             ksFile = Files.newInputStream(Paths.get(keyStoreFilePath));
             LOG.info("KeyStore found on filesystem on path {}", keyStoreFilePath);
         } catch (NoSuchFileException e) {
+            LOG.info("KeyStore not found on filesystem, looking in resources on path {}", keyStoreFilePath);
             LOG.info("KeyStore not found on filesystem, looking in resources on path {}", keyStoreFilePath);
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java:227]: Duplicate logging message.

**Review by ihrasko — CHANGES_REQUESTED** [review]

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java (hunk 6 above)
Comment (ihrasko) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java:27]: OK, go with constructor like in parent. But d not create instance in constructor. And yes, you can add `withSecurityConfig(SecurityConfig)`.

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java (hunk 5 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java]: done

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java (hunk 6 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java]: Done

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java (hunk 7 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java:227]: Fixed

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java [hunk 8]
```diff
@@ -10,30 +10,120 @@
...
     @Test
-    public void rncLightyModuleSmokeTest() throws ConfigurationException {
-        final RncLightyModule rncModule = new RncLightyModule(RncLightyModuleConfigUtils.loadDefaultConfig(),
-                MODULE_TIMEOUT);
-        rncModule.initModules();
-        rncModule.close();
+    public void rncLightyModuleDefaultConfigTest() throws Exception {
+        final var rncModule = new RncLightyModule(RncLightyModuleConfigUtils.loadDefaultConfig(), MODULE_TIMEOUT);
+        assertTrue(rncModule.initModules());
+        final var http1Client = getHttp1Client();
```
Comment (ihrasko) [inline lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java:45]: Unnecessary variable `http1Client`.

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/main/java/io/lighty/applications/rnc/module/RncLightyModule.java [hunk 9]
```diff
@@ -115,21 +115,25 @@ private NetconfSBPlugin initNetconf(final NetconfConfiguration config, final Lig
...
+            this.jettyServerBuilder = new HttpsLightyServerBuilder(inetSocketAddress)
+                    .addSecurityConfig(serverConfig.getSecurityConfig())
+                    .withDefaultJettyServer();
         } else {
             this.jettyServerBuilder = new LightyServerBuilder(inetSocketAddress);
         }
 
         return CommunityRestConfBuilder.from(restConfConfiguration)
-            .withLightyServer(jettyServerBuilder)
+            .withLightyServer(this.jettyServerBuilder)
```
Comment (ihrasko) [inline lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/main/java/io/lighty/applications/rnc/module/RncLightyModule.java:136]: Unnecessary this.

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/main/java/io/lighty/applications/rnc/module/RncLightyModule.java [hunk 10]
```diff
@@ -206,8 +204,7 @@ public boolean close() {
     private boolean stopAndWaitLightyModule(final LightyModule lightyModule) {
         try {
             LOG.info("Stopping lighty.io module ({})...", lightyModule.getClass());
-            final boolean stopSuccess =
-                    lightyModule.shutdown().get(lightyModuleTimeout, DEFAULT_LIGHTY_MODULE_TIME_UNIT);
+            final var stopSuccess = lightyModule.shutdown().get(lightyModuleTimeout, DEFAULT_LIGHTY_MODULE_TIME_UNIT);
```
Comment (ihrasko) [inline lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/main/java/io/lighty/applications/rnc/module/RncLightyModule.java:207]: Unnecessary changes.

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java [hunk 11]
```diff
@@ -10,30 +10,120 @@
...
+    }
+
+    private HttpRequest createGetRequest(final String path) {
+        return HttpRequest.newBuilder()
+                .uri(URI.create(path))
+                .timeout(Duration.ofMinutes(1))
+                .build();
+    }
+
+    private TrustManager[] getTrustedAll() {
```
Comment (ihrasko) [inline lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java:95]: `private static`

Isn't there anything already prepared which accepts all certificates?

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java [hunk 12]
```diff
@@ -10,30 +10,120 @@
...
-        final Boolean isStarted = rncModule.initModules();
+        final var rncModule = new RncLightyModule(config, MODULE_TIMEOUT);
+        final var isStarted = rncModule.initModules();
+        final var isClose = rncModule.close();
 
         assertFalse(isStarted);
+        assertTrue(isClose);
+    }
+
+    private HttpClient getHttp2Client() throws Exception {
```
Comment (ihrasko) [inline lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java:76]: `private static`

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java [hunk 13]
```diff
@@ -10,30 +10,120 @@
...
+        final var ssl = SSLContext.getInstance("SSL");
+        ssl.init(null, getTrustedAll(), new SecureRandom());
+        return HttpClient.newBuilder()
+                .sslContext(ssl)
+                .build();
+    }
+
+    private HttpClient getHttp1Client() {
```
Comment (ihrasko) [inline lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java:84]: Do we really need this one line method?

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/resources/config.json [hunk 14]
```diff
@@ -50,8 +50,9 @@
         "jsonRestconfServiceType": "DRAFT_02"
     },
     "lighty-server" : {
-      "useHttps": false,
+      "useHttp2": true,
```
Comment (ihrasko) [inline lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/resources/config.json:53]: Do we have test for `false` case?

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/HttpsLightyServerBuilder.java [hunk 15]
```diff
@@ -15,36 +16,36 @@
...
+        return this;
     }
 
     @Override
     public Server build() {
-        final Server server = super.build();
-        final SslConnectionFactory ssl = securityConfig.getSslConnectionFactory(HttpVersion.HTTP_1_1.asString());
-        final ServerConnector sslConnector = new ServerConnector(server,
-                ssl, httpConfiguration(this.inetSocketAddress));
+        Preconditions.checkState(securityConfig != null);
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/HttpsLightyServerBuilder.java:34]: How we can enable unsecure connection now? When security config is mandatory?

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java [hunk 16]
```diff
@@ -69,48 +96,136 @@ public static LightyServerConfig getDefaultLightyServerConfig() throws Configura
...
-                throw new ConfigurationException("Unable to create KeyStore configuration: KeyStore file was not found"
-                        + " on path: " + config.getKeyStoreFilePath());
-            }
+            final KeyStore keyStore = getKeyStore(isDefaultKeystoreConfig(config), config.getKeyStoreFilePath(),
+                    config.getKeyStoreType(), config.getKeyStorePassword());
+            LOG.info("Keystore successfully loaded");
+            final KeyStore trustKeyStore = getKeyStore(isDefaultTrustedKeystoreConfig(config),
+                    config.getTrustKeyStoreFilePath(), config.getKeyStoreType(), config.getTrustKeyStorePassword());
+            LOG.info("Trust keystore successfully loaded");
+            LOG.debug("Creating Security config from keystore [{}] and trustKeystore [{}]", keyStore, trustKeyStore);
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java:105]: keystore = `keystore` trustkeystore = `trustore`

**Review by ihrasko — CHANGES_REQUESTED** [review]

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/HttpsLightyServerBuilder.java (hunk 15 above)
Comment (ihrasko) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/HttpsLightyServerBuilder.java:34]: Probably it was better to set security config in constructor if it is required..

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/config/SecurityConfig.java [hunk 17]
```diff
@@ -18,15 +18,17 @@ public class SecurityConfig {
...
 
     public SecurityConfig(final KeyStore keyStore, final String ksPassword, final KeyStore trustKeyStore,
-                          final String trustKsPassword, final boolean isNeedClientAuth) {
+                          final String trustKsPassword, final boolean isNeedClientAuth, final boolean isEnabledSNI) {
         this.keyStore = keyStore;
         this.ksPassword = ksPassword;
         this.trustKeyStore = trustKeyStore;
         this.trustKsPassword = trustKsPassword;
         this.isNeedClientAuth = isNeedClientAuth;
+        this.enableSNI = isEnabledSNI;
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/config/SecurityConfig.java:31]: Please use the same name `enabledSNI`.

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java [hunk 18]
```diff
@@ -69,48 +96,136 @@ public static LightyServerConfig getDefaultLightyServerConfig() throws Configura
...
+            return new SecurityConfig(keyStore, config.getKeyStorePassword(), trustKeyStore,
+                    config.getTrustKeyStorePassword(), config.isEnableSniHostCheck(), config.isNeedClientAuth());
+        } catch (IOException | NoSuchAlgorithmException | CertificateException | KeyStoreException
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java:108]: `final`

File: lighty-modules/lighty-jetty-server/src/test/java/io/lighty/server/LightyServerBuilderTest.java [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+        final var serverBuilder = new HttpsLightyServerBuilder(new InetSocketAddress(PORT))
+                .addSecurityConfig(lightyServerConfig.getSecurityConfig())
+                .withDefaultJettyServer();
+        final var server = initLightyServer(serverBuilder);
+        assertNotNull(server);
+    }
+
+
+    private Server initLightyServer(final LightyServerBuilder serverBuilder) {
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/test/java/io/lighty/server/LightyServerBuilderTest.java:80]: `private static`

File: lighty-modules/lighty-jetty-server/src/test/resources/http2Config.json [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+  "lighty-server": {
+    "useHttp2": true,
+    "keyStorePassword": "ksPassword",
+    "keyStoreType": "JKS",
+    "enableSniHostCheck": true,
+    "keyStoreFilePath": "keystore/keyStore.jks",
+    "trustKeyStorePassword": "trustKsPassword",
+    "trustKeyStoreFilePath": "keystore/trustedKeyStore.jks"
+  }
+}
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/test/resources/http2Config.json:11]: No empty line EOF.

File: lighty-modules/lighty-jetty-server/src/test/resources/httpsConfig.json [hunk 21]
```diff
@@ -0,0 +1,7 @@
+{
+  "lighty-server": {
+    "useHttps": true,
+    "needClientAuth": true,
+    "enableSwagger": true
+  }
+}
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/test/resources/httpsConfig.json:7]: No empty line EOF.

**Review by ihrasko — CHANGES_REQUESTED** [review]

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        httpsConfig.setSecureScheme("https");
+        httpsConfig.setSecurePort(this.inetSocketAddress.getPort());
+        httpsConfig.setSendXPoweredBy(true);
+        httpsConfig.setSendServerVersion(true);
+        httpsConfig.addCustomizer(new SecureRequestCustomizer(securityConfig.isEnableSNI()));
+
+        // HTTP/2 Connection Factory
+        final var h2 = new HTTP2ServerConnectionFactory(httpsConfig);
+        final var alpn = new ALPNServerConnectionFactory();
+        alpn.setDefaultProtocol(HttpVersion.HTTP_1_1.asString());
```
Comment (ihrasko) [inline lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java:50]: `HTTP_2` :D

**Review by ihrasko — CHANGES_REQUESTED** [review]

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java (hunk 8 above)
Comment (PeterSuna) [reply lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java]: I like it and I would like to keep this variable as it describes that this test use HTTP1 client

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/main/java/io/lighty/applications/rnc/module/RncLightyModule.java (hunk 10 above)
Comment (PeterSuna) [reply lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/main/java/io/lighty/applications/rnc/module/RncLightyModule.java]: Removed

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java (hunk 11 above)
Comment (PeterSuna) [reply lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java]: Fixed. No, at least I don't know about that.

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java (hunk 12 above)
Comment (PeterSuna) [reply lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java]: Done

File: lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java (hunk 13 above)
Comment (PeterSuna) [reply lighty-applications/lighty-rnc-app-aggregator/lighty-rnc-module/src/test/java/io/lighty/applications/rnc/module/RncLightyModuleSmokeTest.java]: Removed

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/config/SecurityConfig.java (hunk 17 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/config/SecurityConfig.java]: Done

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java (hunk 18 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java]: Done

File: lighty-modules/lighty-jetty-server/src/test/java/io/lighty/server/LightyServerBuilderTest.java (hunk 19 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/test/java/io/lighty/server/LightyServerBuilderTest.java]: done

File: lighty-modules/lighty-jetty-server/src/test/resources/http2Config.json (hunk 20 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/test/resources/http2Config.json]: Done

File: lighty-modules/lighty-jetty-server/src/test/resources/httpsConfig.json (hunk 21 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/test/resources/httpsConfig.json]: Done

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java (hunk 22 above)
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/Http2LightyServerBuilder.java]: Fixed

File: lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java [hunk 23]
```diff
@@ -57,60 +81,147 @@ public static LightyServerConfig getServerConfiguration(final InputStream jsonCo
...
+                && config.getKeyStorePassword().equals(defaultConfig.getKeyStorePassword())
+                && config.getKeyStoreType().equals(defaultConfig.getKeyStoreType());
+    }
+
+    private static KeyStore getCustomKeystore(final String ksPath, final String ksType, final String ksPassword)
+            throws ConfigurationException, IOException, KeyStoreException, CertificateException,
+            NoSuchAlgorithmException {
+        LOG.info("Loading custom keystore at path : {}", ksPath);
+        final var passProtection = new PasswordProtection(ksPassword.toCharArray());
+        final var keystore = Builder.newInstance(ksType, null, passProtection).getKeyStore();
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-modules/lighty-jetty-server/src/main/java/io/lighty/server/util/LightyServerConfigUtils.java:205]: ignore

**Review by ihrasko — APPROVED** [review]
